### ISSUE TITLE
Allow NBIO paths to have qualifier prefixes

### ISSUE THREAD

**Issue by jshook** [nosqlbench/nosqlbench#375]
**What is the ideal scenario for using this feature?**
A user specifies a local file, classpath resource, or a URL, and they want to be able to specific which of these locations to limit the search in, to avoid ambiguity about the source of the content.
The formats suggested are:
- `file:...` should limit the search to the default filesystem
- `cp:...` or `classpath:...` should search the classpath
- `local:...` should search both classpath and the filesystem
- `all:...` should provide the default behavior for cases where an explicitly selected behavior is needed for scripting integration and so on.
- URLs are already handled with protocol-specific handlers, and do not need separate qualifiers, as these usage scenarios are mutually exclusive. (The `all:...` scenario encapsulates this)

Further, a format of `cp,file:...` may be used to specific an alternate order of precedence.

**Is your feature request related to a problem? Please describe.**
When users have the same-named file in multiple locations, the order of precedence is not obvious.

**Would you like to help implementing or testing it?**

**Additional context**

**Comment by dave2wave** [comment]
Why reinvent uri protocols? I asked ChatGPT about a http+file: and here is the reply:

The URI (Uniform Resource Identifier) protocol field does not have a standard syntax for specifying an "either/or" approach such as `http+file:`. The URI scheme is designed to indicate a single protocol or access method to be used. Here are a few points about the structure and use of URIs:

1. **Standard URI Schemes**: Each URI scheme (like `http`, `https`, `ftp`, `file`, etc.) specifies a single protocol. For example, `http://example.com` specifies the HTTP protocol, and `file:///path/to/file` specifies the file protocol.

2. **Combining Protocols**: There isn't a standard way to combine multiple protocols in a single URI scheme. The idea of `http+file:` is not recognized by any standard and would not be understood by web browsers or other URI processors.

3. **Custom Schemes**: While creating custom URI schemes (like `myapp:`) is possible, they require custom handlers to interpret them. They are not part of any official standard and would need to be explicitly supported by the application using them.

4. **Alternative Approaches**:
    - **Conditional Logic in Applications**: Instead of combining protocols in the URI, handle the protocol decision within the application logic. For instance, your application could try accessing a resource via `http` first and fall back to `file` if the HTTP request fails.
    - **Use of Proxy**: In some cases, a proxy or intermediary service could be used to determine the appropriate protocol to use based on availability or other criteria.

5. **URI Prefixes**: Some applications might use a prefix system to differentiate resource locations, but this is application-specific and not a URI standard. For example, a custom application might interpret `prefix+resource://example` where `prefix` indicates a specific handling mechanism.

In summary, there is no standard for an "either/or" approach in the URI scheme itself. Protocol selection logic is typically handled within the application using the URI rather than within the URI syntax.

**Comment by dave2wave** [comment]
Add Advisor on potential conflicts.

**Comment by dave2wave** [comment]
Focusing on this issue. It looks like this can be completely covered in `NBIO.java`. We just need to carefully preprocess the filename in all of the `read*` methods which currently NBIO.all();

**Comment by jshook** [comment]
> Why reinvent uri protocols? I asked ChatGPT about a http+file: and here is the reply:

The goal here is to make it easy to be the least specific where it is obvious but allow users to be pedantic when needed. And I believe you've seen where URIs are used internally to normalize resource access. So we do use it, but we don't force users to be overly verbose around basic paths when it is otherwise obvious.

**Comment by dave2wave** [comment]
> The goal here is to make it easy to be the least specific where it is obvious but allow users to be pedantic when needed. And I believe you've seen where URIs are used internally to normalize resource access. So we do use it, but we don't force users to be overly verbose around basic paths when it is otherwise obvious.

classpath for example is a reason to do `cp:` as a shorthand.

```
2024-11-22 22:06:00,589 [main] INFO : setting param alias=cp:qdrant_vectors_live.yaml
2024-11-22 22:06:00,594 [main] INFO : workload URI: 'jar:file:///Users/dave.fisher/nb5.jar!/activities/qdrant_vectors_live.yaml'
2024-11-22 22:06:00,623 [main] INFO : workload URI: 'jar:file:///Users/dave.fisher/nb5.jar!/activities/qdrant_vectors_live.yaml'
```
